Add API for changing the context of a running task.

Fuyukai opened the pull request:
This is implemented as an asynchronous context manager (allowing arbitrary nesting).

smurfix commented:
Hmm, I'd call this function `changed_context` so that the `with` statement reads like a reasonable English phrase.

Fuyukai commented:
That's inconsistent with every other context manager in the library.

goodboy reviewed (approved):
Looks fine to me; any lower level testing is just testing `contextvars` itself 😂 

I wonder would example of usage in a real `trio` example be of note?

oremanj reviewed (changes requested):
Thanks for the change! I have some mixed feelings about the "I'm going to commit this regardless of review once it's been up for a week" -- I think based on typical review habits in this project that a week of silence doesn't actually constitute a mandate to go ahead. I can't deny it prompted me to review ASAP, but it felt a little coercive? I'd be more open to this sort of thing if we agree on a standard.

Fuyukai commented on trio/__init__.py:
```diff
@@ -32,6 +32,7 @@
     BrokenResourceError,
     EndOfChannel,
     Nursery,
+    change_context,
```
I was very unsure if this should be in public or lowlevel, I basically did a mental coin flip. I'll change it back.

Fuyukai commented on trio/_core/_context.py:
```diff
@@ -0,0 +1,24 @@
+        yield
+    finally:
+        task.context = saved_context
+        # I assume this is right?
```
The second one is cancel-shielded because it would produce a second cancellation event if the stuff inside was cancelled. The first one isn't because it seems better to not have to do two round trips through the scheduler if the task is already cancelled.

oremanj replied:
That's fair -- I guess anyone who wants to change the context without taking a cancellation can also use a shielded scope. I think most permutations here are defensible as long as they're documented.

Fuyukai commented on docs/source/reference-core.rst:
```diff
@@ -1020,6 +1020,12 @@ Example output (yours may differ slightly):
    request 0: Helper task b finished
    request 0: Request received finished
 
+You can change the current :class:`contextvars.Context` a task is running
```
Okay, I'll move it to the low-level docs as its own section.

Fuyukai commented:
> I have some mixed feelings about the "I'm going to commit this regardless of review once it's been up for a week" -- I think based on typical review habits in this project that a week of silence doesn't actually constitute a mandate to go ahead.

This feature is blocking the project I'm working on. Given that it's a ten line feature change, relatively simple, and that the review was primarily about semantics I think that it's fair to expect this could be pushed through quickly.

oremanj commented:
> This feature is blocking the project I'm working on. Given that it's a ten line feature change, relatively simple, and that the review was primarily about semantics I think that it's fair to expect this could be pushed through quickly.

You could inline `change_context()` into your project in the meantime -- it doesn't actually use any private APIs. I'm not trying to pass judgment on whether pushing on this change's timeline was abstractly fair or not; mostly wanted to flag that I was personally somewhat taken aback by the approach.

pquentin commented:
@Fuyukai any progress on this?

Fuyukai commented:
Hi, I picked this back up again after ~five months of being a writer :)

I addressed all the comments raised by @oremanj. I also wish to apologise for my rash behaviour and threatening a unilateral merge.
